pkg/controller/jobframework/reconciler.go

```diff
@@ -1148,22 +1150,72 @@ func (r *JobReconciler) ignoreUnretryableError(log logr.Logger, err error) error
 	return err
 }
 
-func generatePodsReadyCondition(job GenericJob, wl *kueue.Workload) metav1.Condition {
-	conditionStatus := metav1.ConditionFalse
-	message := "Not all pods are ready or succeeded"
-	// Once PodsReady=True it stays as long as the workload remains admitted to
-	// avoid unnecessary flickering the condition when the pods transition
-	// from Ready to Completed. As pods finish, they transition first into the
-	// uncountedTerminatedPods staging area, before passing to the
-	// succeeded/failed counters.
-	if workload.IsAdmitted(wl) && (job.PodsReady() || apimeta.IsStatusConditionTrue(wl.Status.Conditions, kueue.WorkloadPodsReady)) {
-		conditionStatus = metav1.ConditionTrue
-		message = "All pods were ready or succeeded since the workload admission"
+func generatePodsReadyCondition(log logr.Logger, job GenericJob, wl *kueue.Workload) metav1.Condition {
+	status := metav1.ConditionFalse
+	message := "Workload isn't admitted"
+	reason := kueue.WorkloadWaitForPodsStart
+	if !workload.IsAdmitted(wl) {
+		// The workload has not been admitted yet
+		// or it was admitted in the past but it's evicted/requeued
+		return metav1.Condition{
+			Type:    kueue.WorkloadPodsReady,
+			Status:  status,
+			Reason:  reason,
+			Message: message,
+			// ObservedGeneration is added via workload.UpdateStatus
+		}
+	}
+
+	podsReadyStartCond := workload.IsPodsReadyStartCond(wl)
+	podsReadyRecoveryCond := workload.IsPodsReadyRecoveryCond(wl)
+	log.V(2).Info("PATRYK before switch",
+		"WaitForPodsReady with WorkloadWaitForPodsStart reason:", podsReadyStartCond,
+		"WaitForPodsReady with WorkloadWaitForPodsRecovery reason:", podsReadyRecoveryCond,
+		"Workload admitted:", workload.IsAdmitted(wl),
+		"Pods are ready:", job.PodsReady())
+	switch {
+	// 1a. The workload has just been admitted and doesn't have PodsReady cond yet
+	// 1b. The workload has just been admitted, and not all pods are ready
+	case !job.PodsReady() && (apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadPodsReady) == nil ||
```

Comment:
Make sure PodsReady is computed only once in this function. While it is fast for batch/Job it is slower for JobSet (where it depends on the number of PodSets), but in general it depends on the integration code, so could be non-constant.

Reply:
Good point, done